lib/Dialect/SV/Transforms/AlwaysFusion.cpp

```diff
@@ -0,0 +1,131 @@
+//===- AlwaysFusion.cpp - Always Fusion Pass ------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This transformation pass performs simple fusion of always_ff in the same
+// region.
+//
+//===----------------------------------------------------------------------===//
+
+#include "SVPassDetail.h"
+#include "circt/Dialect/SV/SVDialect.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "mlir/IR/Visitors.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace sv;
+
+namespace {
+
+/// Check the equivalence of operations by doing a deep comparison of operands
+/// and attributes, but does not compare the content of any regions attached to
+/// each op.
+struct SimpleOperationInfo : public llvm::DenseMapInfo<Operation *> {
+  static unsigned getHashValue(const Operation *opC) {
+    return OperationEquivalence::computeHash(const_cast<Operation *>(opC));
+  }
+  static bool isEqual(const Operation *lhsC, const Operation *rhsC) {
+    auto *lhs = const_cast<Operation *>(lhsC);
+    auto *rhs = const_cast<Operation *>(rhsC);
+    if (lhs == rhs)
+      return true;
+    if (lhs == getTombstoneKey() || lhs == getEmptyKey() ||
+        rhs == getTombstoneKey() || rhs == getEmptyKey())
+      return false;
+    return OperationEquivalence::isEquivalentTo(const_cast<Operation *>(lhsC),
+                                                const_cast<Operation *>(rhsC));
+  }
+};
+
+// Merge two regions together. These regions must only have a one block.
+static void mergeRegions(Region *region1, Region *region2) {
+  assert(region1->getBlocks().size() <= 1 && region2->getBlocks().size() <= 1 &&
+         "Can only merge regions with a single block");
+  if (region1->empty()) {
+    // If both regions are empty, move on to the next pair of regions
+    if (region2->empty())
+      return;
+    // If the first region has no block, move the second region's block over.
+    region1->getBlocks().splice(region1->end(), region2->getBlocks());
+    return;
+  }
+  // If the second region is not empty, splice its block into the end of the
+  // first region.
+  if (!region2->empty()) {
+    auto &block1 = region1->front();
+    auto &block2 = region2->front();
+    // Remove the terminator from the first block before merging.
+    block1.back().erase();
```

Comment:
You may want an assert here to make sure the last op is the expected terminator. This adds a bit of future proofing if something changes with always_ff.

Reply:
Good idea! Assert added